### ISSUE TITLE
Help regarding deep kernel learning with graph networks

### ISSUE THREAD

**Issue by arnaghosh** [cornellius-gp/gpytorch#717]
Hi, 
I was trying to combine graph convolutional networks from the DGL library (https://docs.dgl.ai/#get-started) and do graph classification (https://docs.dgl.ai/tutorials/basics/4_batch.html) with a GPR as the regression/classification layer. 
I was stuck on the fact that we need to set the train data and labels before training. I was unable to set the training data as a graph structure, hence I tried setting the initial feature vectors of all nodes (converted to one flattened vector) as the training data. However, when the features are passed through the GCN layers, they are modified and therefore the input to the GPR is different from what was earlier set. This gives errors (rightly so) while training the entire setup together. Is there any workaround for not setting the train_data beforehand?
Any help is much appreciated.
Thanks in advance. :)

**Comment by jacobrgardner** [comment]
Hmm. This is easily fixed for training by turning `gpytorch.settings.debug` off, but you are going to run in to legitimate trouble here when you end up wanting to make predictions with your model, where we do actually use the training data stored in the model.

I have two suggestions: either (1) use a *variational* DKL model if you are planning on using SGD anyways since you'll need to, and the variational models don't require knowing the training data ahead of time, or (2) if you aren't using SGD create a new Module that does something like the following:

```python
class GCNGP(gpytorch.models.GP):
    def __init__(self, gp, gcn, train_x, train_y):
        """
            Args:
                - gp (gpytorch.models.ExactGP): A GP that expects to operate on features extracted by a GCN.
                - gcn (torch.nn.Module): Some PyTorch module that extracts graph features.
                - train_x (any input to gcn): The training data as expected by the GCN
                - train_y (torch.Tensor): Training labels
        """
        self.gp = gp
        self.gcn = gcn
        self.train_x = train_x
        self.train_y = train_y
    
    def forward(self, x):
        train_x_features = self.gcn(self.train_x)
        self.gp.set_train_data(train_x_features, self.train_y)

        x_features = self.gcn(x)
        return self.gp(x_features)
```
And then use this GCNGP in place of your exact GP.

**Comment by arnaghosh** [comment]
Hi Jake,
Thank you so much for your reply. 
The first solution looks interesting but I am currently using Adam. I can try it out and see if the solution works for me. I am hoping that replacing the linear layer with a GP regression layer would improve accuracy. Will surely give it a try.
I was actually trying to do something like your second solution. I did not create a separate module but was trying to call the set_train_data function for every training batch forward pass. I am pretty sure this is wrong because I am constantly changing the training data in every batch. So when I go to evaluation mode, it gives me an error (probably because the number of training samples is smaller than validation samples). 
As most of the processing in graph networks is done in mini-batches, could you suggest me a way to do what I want without calling the gcn_forward on self.train_x everytime (1st line of the forward function posted by you)? 
Thanks again.

**Comment by arnaghosh** [comment]
Hey,
I was wondering if there was any update on how to achieve this? 
Thanks for the help again. :)

**Comment by jacobrgardner** [comment]
@arnaghosh If you'd like to use minibatch training, you'll need to use a variational DKL model, like the ones we have [here](https://github.com/cornellius-gp/gpytorch/blob/master/examples/05_Scalable_GP_Regression_Multidimensional/SVDKL_Regression_SVGP_CUDA.ipynb).

These will actually be much simpler to adapt to your setting, because as you'll notice the variational model does not assume access to the full training dataset ahead of time.

**Comment by arnaghosh** [comment]
Thanks a lot for your suggestion. I will give it a try and let you know how it goes :)

**Comment by gpleiss** [comment]
closing for now - reopen if you still have questions :)
